### DISCUSSION TITLE
Assistance request:  digest and transfer of useful functionalities from personal repo

### DISCUSSION THREAD

**Discussion by DanNBullock** [dipy/dipy#2620] [1. Ideas & Feedback]
Hi all,

### Overview
Back in the day, I made a repository of[ matlab-based tools (wma_tools)](https://github.com/DanNBullock/wma_tools) as an extension [vistasoft](https://github.com/vistalab/vistasoft)/[mba](https://github.com/francopestilli/mba) functionalities, and (as it turned out--though I didn't know it at the time) a recapitulation of various [WMQL](https://github.com/demianw/tract_querier) capabilities.  Eventually I made the shift to python and translated most of the functionality over to python: [wma_pyTools](https://github.com/DanNBullock/wma_pyTools).  Currently I foresee upcoming changes in my ability to update and continue work on this project, and I was hoping to transfer them to the more sustainable and widely used DIPY project.

### Admitted limitations
To be clear: wma_pyTools is **rough** and **unpolished** and probably in great need of critical eyes.  Thus far it has been a one-person operation, and so I've had to balance my time between creating omnibus documentation, in-code documentation, and actual code development (and all when permitted by other, ongoing work).  As such, many functions may be sub-optimally implemented and/or poorly documented--though I am working to improve this and there tends to be a good deal of in-code commentary for better or worse.  Even so, I think there's a non-trivial amount of stuff that may be useful in there.  

#### Broad overview
For example, there's the WMQL-like functionalities contained in the [roiTools](https://github.com/DanNBullock/wma_pyTools/blob/master/wmaPyTools/roiTools.py) sub-module and various segmentation tools/functions contained in the [segmentationTools](https://github.com/DanNBullock/wma_pyTools/blob/master/wmaPyTools/segmentationTools.py) sub-module.  Overall, the function collections (the toolkit is written in a function-oriented rather than object-oriented fashion) are divided into the following sub-modules:

#### Sub-module list
- [analysisTools](https://github.com/DanNBullock/wma_pyTools/blob/master/wmaPyTools/analysisTools.py):  tools for the derivation and production of quantitative measures from standard imaging and tractography inputs (e.g. streamlines, tractograms, nifti objects).  
- [roiTools](https://github.com/DanNBullock/wma_pyTools/blob/master/wmaPyTools/roiTools.py): tools for the creation, manipulation, and usage of nifti-based regions of interest (ROIS).
- [segmentationTools](https://github.com/DanNBullock/wma_pyTools/blob/master/wmaPyTools/segmentationTools.py):  tools for the application / implementation of tractography segmentations (typically heavily reliant on products from _streamlineTools_ and _roiTools_
- [streamlineTools](https://github.com/DanNBullock/wma_pyTools/blob/master/wmaPyTools/streamlineTools.py): tools for interactions with and manipulations of streamline and tractogram--based objects.
- [visTools] (https://github.com/DanNBullock/wma_pyTools/blob/master/wmaPyTools/visTools.py):  various functions intended to serve as easy to use interfaces for the production of quasi-publication level images. [As a biased aside, some of these functionalities, like the "clean" (i.e. de-cluttered/restructured) version of [_dipyPlotTract_](https://github.com/DanNBullock/wma_pyTools/blob/8d28ac89bf77c2b469a1e9bf666fad99387d62d0/wmaPyTools/visTools.py#L1885-L2027) can produce \*very\* aesthetically appealing figures]
-  [genUtils](https://github.com/DanNBullock/wma_pyTools/blob/master/wmaPyTools/genUtils.py):  general utility functionalities, typically for IO operations and/or as related to brainlife.io interoperability.

#### Specific function(s) example: fast segmentation
As a particular example of one functionality (among many, hopefully) that might be useful (and one which I have discussed to some degree with @arokem and @skoudoro), I managed to speed up the segmentation process by something approximating an order of magnitude in many cases ([as revealed in testing](https://github.com/DanNBullock/wma_pyTools/tree/master/testsOrDemos), though their appears to be a floor effect incurred by some initial bounding-box computations) by implementing a number of heuristics for [near_roi](https://github.com/DanNBullock/wma_pyTools/blob/6cb543b25c8f66658cd0c07a388b1c49c298af87/wmaPyTools/segmentationTools.py#L501-L766) and the ["smart" sequencing](https://github.com/DanNBullock/wma_pyTools/blob/6cb543b25c8f66658cd0c07a388b1c49c298af87/wmaPyTools/segmentationTools.py#L34-L129) of [multiple segmentation criteria](https://github.com/DanNBullock/wma_pyTools/blob/6cb543b25c8f66658cd0c07a388b1c49c298af87/wmaPyTools/segmentationTools.py#L251-L347).

### Specific request(s)
To this end I was hoping that someone(s) might be available to help me with the following tasks:

1. Identifying which functionalities are _both_ consistent with the intended scope of DIPY and worth incorporation.
2. In such cases, identifying requisite changes in comment and coding _style_,  _beyond_ requisite PEP8 changes (the contributing guide appears to leave a great deal open to the developers choice)
3. Also in such cases, identifying requisite _structural_ (e.g. object oriented standards/characteristics, standard output formats/objects) changes that ought to be made.
4. Should time and attention permit: general suggestions / recommendations regarding more optimal implementations and overall reductions in the amount of redundant code (either with respect to within-package functionalities, or functionalities implemented in other imported/importable modules)

#### Final note
If this request/discussion ought to be reformatted/rewritten/triaged/shelved just let me know.

**Comment by skoudoro** [comment]
Hi @DanNBullock,

Thank you for this request. I will be happy to help you with this. 

it seems that ``wma_tools`` have much more features than ``wma_pyTool``?  Am I right? Did you succeed to migrate everything?

One easy solution could be to migrate this project to DIPY organization. Then reshape it, and add this package as an optional dependency for DIPY. It is a way to keep your package alive longer. Then, we can slowly migrate multiple functions.

Will you be interested in this solution? It implies that you will have to be an active maintainer of wma_pytool (which is already the case). Of course, we will help but I am not sure what this sentence `Currently I foresee upcoming changes in my ability to update` implies.

I have time this week if you want to chat about it. Feel free to ping me anytime during the afternoon

**Comment by DanNBullock** [comment]
#### `wma_tools` VS `wma_pyTool`

As for the difference between `wma_tools` and `wma_pyTool` (and whether I migrated everything), I think--for the most part--I managed to migrate the important parts over.  There are two primary exceptions to this: redundant versions and [segmentation implementations](https://github.com/DanNBullock/wma_tools/tree/master/Segmentations).  

As for the redundant versions, I began `wma_tools` really early on in my Ph.D. and so I had not been adequately schooled in the ways of code development by Soichi yet.  As such, you'll notice a lot of `[function_name]_V[X]` going on in there for various reasons.  `wma_tools` was also primarily a one person production, and so I was often very concerned about breaking functioning scripts/code.  Because I didn't necessarily understand git versioning or branches I would instead just append a `_V[x]) on to new versons--hence the mess.

As for segmentation, there's some ongoing discussion about what to do with these.  They are fairly straightforwardly translatable to `wma_pyTools` versions (though somewhat laboriously, if the [correct, human-readable format](https://github.com/DanNBullock/wma_tools/blob/e380162e46b3526ed538c6ff071131f960bb0284/Segmentations/bsc_genericSegmentFunction.m) (e.g. [here, for uncinate](https://github.com/DanNBullock/wma_tools/blob/b0792d8488851bb09423abd606a3ebb45158debd/Segmentations/bsc_segmentAntPostTracts_v4.m#L75-L208)) is used).  We are potentially trying to figure out whether these ought to be incorporated into another project (e.g. [pyAFQ](https://github.com/yeatmanlab/pyAFQ)), that is more focused on _particular_ segmentation implementations, rather than broad, general-purpose toolkits (as seems to be the use case for DIPY).

Generally speaking though, if there happens to be any functionalities that haven't been ported over I'd be more than happy to do so--it's just a matter of me finding such cases or having them brought to my attention.  At some point it may be worthwhile for me to do a systematic review to this effect, but I'm not sure I'll have bandwidth in the near future.

#### Migration to DIPY organization 

I would be extremely happy to migrate `wma_pyTools` over to DIPY.  As you suggest, one of my primary goals with this effort is to ensure the survival and utility of the functionalities I've developed.  To be candid though, I am somewhat hesitant to do this at this exact moment--mostly due to being self-conscious of the current state of my code.  DIPY and its associated projects are professionally maintained and well curated, whereas my work cannot be described in this same way.  There's a non-trivial number of functions without docstrings (though my variable names and comments tend to be painfully and gratingly verbose), which is not great, but there's also a substantial amount of "cruft" in the form of commented out and/or obsolete functions.  Both of these are fairly straightforward issues to fix, they'll just take some dedicated time.  I may want to take a moment to create a branch specifically for this purpose and do some house-cleaning before I pass it over--just for the sake of saving face.

#### Continued maintenance of wma_pyTools

I'm certainly comfortable with serving as the primary maintainer of wma_pyTools, but at the same time, it is my hope to get it to a more stable, production-ready state.  Sure there are functionalities that can be added or optimized, but I think that the active development of new features is likely on a downswing (unless I or someone else comes up with a new suite of worthwhile features).  The bit about upcoming changes in the ability to update reflects an upcoming shift in work roles such that I probably won't be spending _most_ of my time doing active, hands-on code development (e.g. in the capacity of a post-doc), but I'll still likely have time to do code updates and make similar contributions.


I'll send you a message via one of our shared slack channels to discuss this in more detail.  Thanks for the guidance and assistance!

**Comment by arokem** [comment]
Hey @DanNBullock : specifically regarding porting generally-useful functionality to DIPY, I think that it would be the easiest to find things that are relatively self-contained and small and starting with those. For example, `near_roi_fast` would be a good start. IIUC, the API of this function is similar (or identical?) to the API provided by the current implementation of `near_roi`, you could refactor the existing implementation to incorporate your improvements. The benefit of that is that there is already testing in place to cover potential use-cases, and we can see whether the implementation has any issues with these use-cases, and keep fixing from there.

**Comment by DanNBullock** [comment]
@arokem Testing cases was definitely one thing I was thinking about in the back of my head--specifically as it relates to this proposed segmentation update.  I've looked over the [test suite for tracking utils](https://github.com/DanNBullock/dipy/blob/master/dipy/tracking/tests/test_utils.py), and, while I think that `near_roi_fast` would probably pass this test (I can check this on my end), I think that some more complicated tests might be warranted to ensure that:

1.  It's picking out the equivalent / accurate streamlines that `near_roi` is.
2.  There's some sort of report about the _alleged_ speed improvement of `near_roi_fast`, and potentially some check / assurance of the scaling on this (the time-saving increases as a greater number of streamlines are assessed).  Maybe this entails something as simple as a `timeit`, relational operator, and a warning flag?
3.  That the implemented heuristics (and their underlying assumptions) aren't encountering any unforeseen edge cases or streamline-based shenanigans.

The current tests for [near_roi](https://github.com/dipy/dipy/blob/f0775487bf01eee8964f5224e176ec99c891da0e/dipy/tracking/tests/test_utils.py#L337-L432) only cover [three relatively sparsely sampled test streamlines occupying an _extremely_ small volume of space](https://github.com/dipy/dipy/blob/f0775487bf01eee8964f5224e176ec99c891da0e/dipy/tracking/tests/test_utils.py#L339-L346).  I recognize the need to keep tests small, simple, and fast, but I worry that the existing tests aren't sufficiently stringent to ensure that `near_roi_fast` is _actually_ working reliably (though I think the current tests are fine for `near_roi` as it isn't making as many assumptions about input streamlines). For example, I have strong reason to expect that [streamline compression, as currently being implemented in the TRX format, will probably break some of the heuristics](https://github.com/tee-ar-ex/trx-python/issues/17)--though, contrary to the aside in the previous sentence, I suspect `near_roi` would be impacted in similar ways.

Is there a way to incorporate a "more naturalistic" test collection of streamlines / tractogram for this purpose (e.g. maybe a standard test tractogram lurking somewhere in the dipy code)?  As I mentioned earlier in this thread, [I was developing a test suite](https://github.com/DanNBullock/wma_pyTools/tree/master/testsOrDemos) to ensure the various characteristics in the above enumerated list, but even if I were to revisit and update those, it seems that such tests are beyond the standard scope of DIPY's testing suite--but I could be wrong about this.
